lib/derivativo/user_path_helper.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Derivativo
+  class UserPathHelper
+    # Checks the user's $PATH for the given program
+    def self.which(program)
+      exts = pathext_values
+      path_values.each do |path|
+        exts.each do |ext|
+          # Do a case-insensitive search (e.g. because windows stores PATHEXT extensions as
+          # capitalized strings and these won't match the on-disk versions of executables).
+          # Note that it's possible to have case-sensitive volumes (or directories) on both
+          # windows and unix-like systems (though extension matching is less likely to be a
+          # problem on unix-like systems.
+          possible_executable_file = case_sensitive_file_path(path, "#{program}#{ext}")
+          return possible_executable_file if File.executable? possible_executable_file
```

Comment:
File.executable? expects a string and will complain when it tries to convert nil (returned by case_sensitive_file_path if no matching file found) into a String (no implicit conversion).
Possible solution: case_sensitive_file_path returns '' instead of nil.

Reply:
You're right about `File.executable?(nil)` normally complaining about not getting a string (in irb or the rails console), but oddly enough I'm finding that this doesn't raise an error in an rspec test:

```
puts 'exist? ' + File.executable?(nil).inspect
```
And it prints out `exist? false`.

Not sure why.

Reply:
I also updated some tests, since your comment also revealed a gap in the testing.